File: 2018/07_TheSumofItsParts/aoc_07.go
// ======================================================================
// The Sum of Its Parts
//   Advent of Code 2018 Day 07 -- Eric Wastl -- https://adventofcode.com
//
// lua implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           a o c _ 0 7 . g o
// ======================================================================
// Solve the puzzles for Advent of Code 2018 day 07

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
    "flag"
    "fmt"
    "os"
    "regexp"
    "strings"
)

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const (
     SAVE_BLANK_LINES = false
)

// ----------------------------------------------------------------------
//                                                                globals
// ----------------------------------------------------------------------
var aocVerbose bool
var aocPart int
var aocLimit int
var aocInput string = "";

// ----------------------------------------------------------------------
//                                                          initialzation
// ----------------------------------------------------------------------

func init() {
  // Parse the command line options

  // 1. Create the command line parser
  flag.IntVar(&aocPart, "part", 1, "Part 1 or 2 of the puzzle")
  flag.IntVar(&aocPart, "p", 1, "Part 1 or 2 of the puzzle")
  flag.IntVar(&aocLimit, "limit", 0, "Optional execution limit")
  flag.IntVar(&aocLimit, "l", 0, "Optional execution limit")
  flag.BoolVar(&aocVerbose, "verbose", false, "Optional extra messages")
  flag.BoolVar(&aocVerbose, "v", false, "Optional extra messages")
  flag.StringVar(&aocInput, "input", "", "Input file name")
  flag.StringVar(&aocInput, "i", "", "Input file name")
}

// ----------------------------------------------------------------------
//                                                       parseCommandLine
// ----------------------------------------------------------------------

func parseCommandLine() {

    // 1. Parse the command line to aocXXXX variables
    flag.Parse()

    // 2. Verify that part is 1 or 2
    if aocPart != 1 && aocPart != 2 {
        fmt.Println("part (-p) must be 1 or 2")
        os.Exit(1)
    }

    // 3. If input file name not specified, get from unprocessed arguments
    if aocInput == "" {
        if flag.NArg() == 0 {
            fmt.Println("Input file name (-i) not specified")
            os.Exit(1)
        }
        aocInput = flag.Arg(0)
    }
}

// ----------------------------------------------------------------------
//                                                                partOne
// ----------------------------------------------------------------------

func partOne(verbose bool, limit int, input_lines []string) string {
    // Process part one of the puzzle

    // 1. Create the puzzle solver
    solver := NewSolver(false, input_lines)

    // 2. Determine the solution for part one
    solution := solver.PartOne(verbose, limit)
    if solution == "" {
        fmt.Println("There is no solution")
    } else {
        fmt.Println("The solution for part one is " + solution)
    }

    // 3. Return result
    return solution
}

// ----------------------------------------------------------------------
//                                                                partTwo
// ----------------------------------------------------------------------

func partTwo(verbose bool, limit int, input_lines []string) string {
    // Process part two of the puzzle

    // 1. Create the puzzle solver
    solver := NewSolver(true, input_lines)

    // 2. Determine the solution for part two
    solution := solver.PartTwo(verbose, limit)
    if solution == "" {
        fmt.Println("There is no solution")
    } else {
        fmt.Println("The solution for part two is " + solution)
    }

    // 3. Return result
    return solution
}

// ----------------------------------------------------------------------
//                                                               fromFile
// ----------------------------------------------------------------------

func fromFile(filepath string) []string {
    // 1. Read the file
    b, err := os.ReadFile(filepath)
    if err != nil {
        fmt.Print(err)
        os.Exit(2)
    }

    // 2. Return the contents as a string
    return fromText(string(b))
}

// ----------------------------------------------------------------------
//                                                               fromText
// ----------------------------------------------------------------------

func fromText(text string) []string {
    // Break the text into trimed, non-comment lines

    // 1. We start with no lines
    var result []string

    // 2. Break up the text into lines
    lines := regexp.MustCompile("\\r?\\n").Split(text, -1)

    // 3. Loop for lines in the text
    for _, line := range lines {

        // 4. Ignore comment lines and maybe blank lines
        if len(line) > 0 {
            if !strings.HasPrefix(line, "!") {
                result = append(result, line)
            }
        } else if SAVE_BLANK_LINES {
            result = append(result, "")
        }
    }

    // 5. Return a table of cleaned text lines
    return result
}

// ----------------------------------------------------------------------
//                                                                   main
// ----------------------------------------------------------------------

func main() {
    // Read the Advent of Code problem and solve it

    // 1. Get the command line options
    parseCommandLine()

    // 2. Read the puzzle file
    input_text := fromFile(aocInput)

    // 3. Process the appropiate part of the puzzle
    var result string
    if aocPart == 1 {
        result = partOne(aocVerbose, aocLimit, input_text)
    } else {
        result = partTwo(aocVerbose, aocLimit, input_text)
    }

    // 5. Set return code (0 if solution found, 2 if not)
    if len(result) > 0 {
        os.Exit(0)
    }
    os.Exit(2)
}

// ======================================================================
// end                         a o c _ 0 7 . g o                      end
// ======================================================================

File: 2024/07_BridgeRepair/aoc_07.py

# ======================================================================
# Bridge Repair
#   Advent of Code 2024 Day 07 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                           a o c _ 0 7 . p y
# ======================================================================
"Solve the puzzles for Advent of Code 2024 day 07"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import argparse
import sys

import equations

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
KEEP_BLANK = False
IGNORE_COMMENTS = False
COMMENT = "!"

# ----------------------------------------------------------------------
#                                                     parse_command_line
# ----------------------------------------------------------------------


def parse_command_line():
    "Parse the command line options"

    # 1. Create the command line parser
    desc = 'Bridge Repair - Day 07 of Advent of Code 2024'
    sample = 'sample: python aoc_07.py input.txt'
    parser = argparse.ArgumentParser(description=desc,
                                     epilog=sample)
    parser.add_argument('-v', '--verbose', action='store_true',
                        default=False, dest='verbose',
                        help='Print status messages to stdout')
    parser.add_argument('-p', '--part', action='store', default=1, type=int,
                        dest='part', help='Puzzle Part (1 or 2)')
    parser.add_argument('-l', '--limit', action='store', default=0, type=int,
                        dest='limit',
                        help='Maximum (time, size, depth) before stopping')
    parser.add_argument('filepath', metavar='FILENAME',
                        action='store', type=str,
                        help="Location of puzzle input")

    # 2. Get the options and arguments
    return parser.parse_args()

# ----------------------------------------------------------------------
#                                                               part_one
# ----------------------------------------------------------------------


def part_one(args, input_lines):  # pylint: disable=E1128
    "Process part one of the puzzle"

    # 1. Create the puzzle solver
    my_solver = equations.Equations(part2=False, text=input_lines)

    # 2. Determine the solution for part one
    solution = my_solver.part_one(verbose=args.verbose, limit=args.limit)
    if solution is None:
        print("There is no solution")
    else:
        print(f"The solution for part one is {solution}")

    # 3. Return result
    return solution is not None

# ----------------------------------------------------------------------
#                                                               part_two
# ----------------------------------------------------------------------


def part_two(args, input_lines):  # pylint: disable=E1128
    "Process part two of the puzzle"

    # 1. Create the puzzle solver
    my_solver = equations.Equations(part2=True, text=input_lines)

    # 2. Determine the solution for part two
    solution = my_solver.part_two(verbose=args.verbose, limit=args.limit)
    if solution is None:
        print("There is no solution")
    else:
        print(f"The solution for part two is {solution}")

    # 3. Return result
    return solution is not None

# ----------------------------------------------------------------------
#                                                              from_file
# ----------------------------------------------------------------------


def from_file(filepath):
    "Read the file"

    return from_text(open(filepath, encoding="utf8").read())

# ----------------------------------------------------------------------
#                                                              from_text
# ----------------------------------------------------------------------


def from_text(text):
    "Break the text into trimed, non-comment lines"

    # 1. We start with no lines
    lines = []

    # 2. Loop for lines in the text
    for line in text.splitlines():

        # 3. But ignore blank and comment lines
        line = line.rstrip()
        if (not line) and (not KEEP_BLANK):
            continue
        if line.startswith(COMMENT) and IGNORE_COMMENTS:
            continue

        # 4. Add the line
        lines.append(line)

    # 5. Return a list of clean lines
    return lines

# ----------------------------------------------------------------------
#                                                                   main
# ----------------------------------------------------------------------


def main():
    "Read the Advent of Code problem and solve it"

    # 1. Get the command line options
    args = parse_command_line()

    # 2. Read the puzzle file
    input_text = from_file(args.filepath)

    # 3. Process the appropiate part of the puzzle
    if args.part == 1:
        result = part_one(args, input_text)
    else:
        result = part_two(args, input_text)

    # 5. Set return code (0 if solution found, 2 if not)
    if result:
        sys.exit(0)
    sys.exit(2)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    main()

# ======================================================================
# end                         a o c _ 0 7 . p y                      end
# ======================================================================
